purchase_transport_multi_address/model/purchase_order.py

```diff
@@ -0,0 +1,123 @@
+# -*- coding: utf-8 -*-
+#    Author: Leonardo Pistone, Yannick Vaucher
+#    Copyright 2014 Camptocamp SA
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from openerp import models, fields, api, exceptions
+from openerp.tools.translate import _
+
+
+class PurchaseOrder(models.Model):
+    _inherit = 'purchase.order'
+
+    origin_address_id = fields.Many2one(
+        'res.partner',
+        'Origin Address')
+    consignee_id = fields.Many2one(
+        'res.partner', 'Consignee',
+        help="The person to whom the shipment is to be delivered.")
+
+    @api.onchange('dest_address_id')
+    def new_onchange_dest_address_id(self):
+        """Find a picking type from the address
+
+        If the address can be an internal warehouse or a customer, the picking
+        type is changed accordingly. This intentionally overrides the original
+        without super, and should be consistent with the module
+        purchase_requisition_delivery_address.
+
+        A similar logic to choose the picking type is used in the module
+        framework_agreement_sourcing in github.com/OCA/vertical-ngo.
+
+        """
+        PickType = self.env['stock.picking.type']
+        types = PickType.search([
+            ('warehouse_id.partner_id', '=', self.dest_address_id.id),
+            ('code', '=', 'incoming'),
+        ])
+
+        if types:
+            if self.picking_type_id in types:
+                return
+            picking_type = types[0]
+        elif self.dest_address_id.customer:
+            # if destination is not for a warehouse address,
+            # we set dropshipping picking type
+            ref = 'stock_dropshipping.picking_type_dropship'
+            picking_type = self.env.ref(ref)
+        else:
+            raise exceptions.Warning(_(
+                'No picking types were found on warehouse. Please verify you '
+                'have set an address on warehouse.'))
+        self.picking_type_id = picking_type
+
+    @api.onchange('picking_type_id')
+    def onchange_picking_type_id(self):
+        """If the picking type has an address, use it.
+
```

Comment:
@nbessi what do you suggest? It would be slightly more accurate to say "If the picking type has a *warehouse with* an address, use it".